oauthenticator/auth0.py

```diff
@@ -0,0 +1,149 @@
+"""
+Custom Authenticator to use Auth0 OAuth with JupyterHub
+
+Derived using the Github and Google OAuthenticator implementations as examples.
+
+The following environment variables may be used for configuration:
+
+    AUTH0_SUBDOMAIN - The subdomain for your Auth0 account
+    OAUTH_CLIENT_ID - Your client id
+    OAUTH_CLIENT_SECRET - Your client secret
+    OAUTH_CALLBACK_URI - Your callback handler URI
+
+Additionally, if you are concerned about your secrets being exposed by
+an env dump(I know I am!) you can create a function that returns them in 
+a dict and attach them to the config object for Auth0OAuthenticator.
+
+One instance of this could be adding the following to your jupyterhub_config.py :
+
+  def oauth_variable_config():
+    return {
+      'OAUTH_CLIENT_ID':'YOUR_CLIENT_ID',
+      'OAUTH_CLIENT_SECRET':'YOUR_CLIENT_SECRET',
+      'OAUTH_CALLBACK_URI':'YOUR_CALLBACK_URI'
+    }
+  c.Auth0OAuthenticator.oauth_variable_config = oauth_variable_config
+
+If you are using the environment variable config, all you should need to
+do is define them in the environment then add the following line to 
+jupyterhub_config.py :
+
+  c.JupyterHub.authenticator_class = 'oauthenticator.auth0.Auth0OAuthenticator'
+
+"""
+
+
+import json
+import os
+
+import logging
+
+from tornado.auth import OAuth2Mixin
+from tornado import gen, web
+
+from tornado.httputil import url_concat
+from tornado.httpclient import HTTPRequest, AsyncHTTPClient
+
+from jupyterhub.auth import LocalAuthenticator
+
+from traitlets import Unicode, Any, default
+
+from .oauth2 import OAuthLoginHandler, OAuthenticator
+
+AUTH0_SUBDOMAIN = os.getenv('AUTH0_SUBDOMAIN')
+
+class Auth0Mixin(OAuth2Mixin):
+    _OAUTH_AUTHORIZE_URL = "https://%s.auth0.com/authorize" % AUTH0_SUBDOMAIN
+    _OAUTH_ACCESS_TOKEN_URL = "https://%s.auth0.com/oauth/token" % AUTH0_SUBDOMAIN
+
+
+class Auth0LoginHandler(OAuthLoginHandler, Auth0Mixin):
+    def get(self):
+        redirect_uri = self.authenticator.oauth_callback_url
+        self.authorize_redirect(
+            redirect_uri=redirect_uri,
+            client_id=self.authenticator.oauth_client_id,
+            scope=self.scope,
+            response_type='code')
+
+class Auth0OAuthenticator(OAuthenticator):
+
+    oauth_variable_config = Any(
+        help="""Any callable that returns a dictionary of oauth variables by the following names:
+        OAUTH_CLIENT_ID
+        OAUTH_CLIENT_SECRET
+        OAUTH_CALLBACK_URI
+        """
+    ).tag(config=True)
+
+    def oauth_environment_variable_config(self):
```

Comment:
Can we skip this, and leave the getting of CLIENT_ID, etc. from the env the same as all the other implementations?